src/OSPSuite.Core/Domain/Populations/ParameterValuesCache.cs

```diff
@@ -0,0 +1,188 @@
+using OSPSuite.Assets;
+using OSPSuite.Core.Extensions;
+using OSPSuite.Utility.Collections;
+using OSPSuite.Utility.Exceptions;
+using OSPSuite.Utility.Extensions;
+
+namespace OSPSuite.Core.Domain.Populations
+{
+   public interface IParameterValueCache
+   {
+      bool Has(string parameterPath);
+      IReadOnlyList<double> ValuesFor(string parameterPath);
+      IReadOnlyList<double> PercentilesFor(string parameterPath);
+      string[] AllParameterPaths();
+
+      /// <summary>
+      ///    Similar to <see cref="ValuesFor" /> but returns an array for R compatibility
+      /// </summary>
+      double[] GetValues(string parameterPath);
+
+      void Remove(string parameterPath);
+      void SetValues(string parameterPath, IReadOnlyList<RandomValue> newValues);
+      void SetValues(string parameterPath, IReadOnlyList<double> newValues);
+      IReadOnlyCollection<ParameterValues> AllParameterValues { get; }
+      ParameterValues ParameterValuesFor(string parameterPath);
+      ParameterValues ParameterValuesAt(int index);
+      void Add(ParameterValues parameterValues);
+      void Add(IReadOnlyCollection<ParameterValue> parameterValues);
+      void RenamePath(string oldPath, string newPath);
+      int Count { get; }
+   }
+
+   /// <summary>
+   ///    Represents a cache containing the population values. Values are stored like so:
+   ///    1 - One column per parameter path
+   ///    2 - All values for one parameter are stored in one column. That means that one individual can be read as a row
+   /// </summary>
+   public class ParameterValuesCache : IParameterValueCache
+   {
+      private readonly ICache<string, ParameterValues> _parameterValuesCache = new Cache<string, ParameterValues>(x => x.ParameterPath);
+
+      public virtual void Add(IReadOnlyCollection<ParameterValue> parameterValues) => addValues(parameterValues);
+
+      public virtual void RenamePath(string oldPath, string newPath)
+      {
+         var possibleKey = internalKeyFor(oldPath);
+         if (string.IsNullOrEmpty(possibleKey))
+            return;
+
+         var values = _parameterValuesCache[possibleKey];
+         Remove(possibleKey);
+         _parameterValuesCache.Add(newPath, values);
+      }
+
+      public virtual int Count
+      {
+         get
+         {
+            if (!_parameterValuesCache.Any())
+               return 0;
+            return _parameterValuesCache.First().Count;
+         }
+      }
+
+      private string internalKeyFor(string parameterPath)
+      {
+         if (_parameterValuesCache.Contains(parameterPath))
+            return parameterPath;
+
+         var allSupportedPath = AllParameterPaths();
+         var index = Array.FindIndex(allSupportedPath, x => x.StartsWith(parameterPath));
+         //try to strip the unit and see if it's possible to find a key exactly equal to the parameterPath without the unit
+         if (index >= 0 && string.Equals(allSupportedPath[index].StripUnit(), parameterPath))
```

Comment:
Checking if we find a key that starts with the full path of the parameter without the unit and verify that they are indeed equal when we strip the unit